### PR TITLE
Update pippy

### PR DESCRIPTION
**Pull request by sparshg**

### CODE REVIEW

**Review by tchx84 — CHANGES_REQUESTED** [review]
Looking good but let's organize the commits the following way:

1. Bump the BaseApp version and GNOME runtime.
2. Update pygame.
3. Update vte.
4. Update x.y.z

Then let's add data-checkers for all the sources. Similarly to what I have done with the BaseApp https://github.com/flathub/org.sugarlabs.BaseApp/tree/branch/23.06

**Review by tchx84 — CHANGES_REQUESTED** [review]
Can you update all the other sources? Plus, don't forget about adding data checkers.

**Review by tchx84 — CHANGES_REQUESTED** [review]
Can you update all the other sources? Plus, don't forget about adding data checkers.

**Comment by sparshg** [comment]
Yep on it, I was rearranging previous commits first.

**Comment by sparshg** [comment]
I have updated all sources except Box2D, it doesn't have sdist on pypi, what should I do?

File: org.sugarlabs.Pippy.json [hunk 1]
```diff
@@ -58,28 +100,39 @@
...
                     "type": "archive",
-                    "url": "https://download.gnome.org/sources/vte/0.60/vte-0.60.1.tar.xz",
-                    "sha256": "5e25807233f8a7e15204be7ff694bbcf6facbb0136092b5ba12584a7b70cf0c4"
+                    "url": "https://download.gnome.org/sources/vte/0.70/vte-0.70.5.tar.xz",
+                    "sha256": "6205fe194c2d800321f0fbb43a282ddf2baa46e8859b5baabbe30f64fd872a24",
+                    "x-checker-data": {
+                        "type": "gnome",
+                        "name": "vte",
+                        "versions": {
+                            "<": "0.71"
```
Comment (tchx84) [inline org.sugarlabs.Pippy.json]: What's the reason for this condition?

**Review by tchx84 — CHANGES_REQUESTED** [review]
A missing things:

1. Removed shared-modules submodule (in a separate commit), since it's no longer used.
2. Add data checker to all sources, including the Pippy git source.
3. Let's streamline commit messages with the `manifest: ` prefix.

**Comment by tchx84** [comment]
> I have updated all sources except Box2D, it doesn't have sdist on pypi, what should I do?

Try https://github.com/flatpak/flatpak-builder-tools/tree/master/pip

File: org.sugarlabs.Pippy.json (hunk 1 above)
Comment (sparshg) [reply org.sugarlabs.Pippy.json]: I remember I had some problem running Pippy with newer versions but its working fine now that I tried again, I will remove this

**Comment by sparshg** [comment]
> Try https://github.com/flatpak/flatpak-builder-tools/tree/master/pip

It did the same thing, so I added the checker anyways. It gives a warning if you check for box2d updates that it couldn't find sdist for box2d.

**Comment by sparshg** [comment]
Okay seems like the bot is failing it due to this box2d error...

**Comment by tchx84** [comment]
> Okay seems like the bot is failing it due to this box2d error...

```
ERROR   src.manifest: Failed to check file box2d/Box2D-2.3.2.tar.gz with PyPIChecker: Couldn't find sdist for package Box2D: list index out of range
```

Please join Flathub channel on Matrix and ask what can be done here.

**Comment by sparshg** [comment]
Can we not replace it with git for the source instead of pypi, https://github.com/pybox2d/pybox2d/releases

**Comment by tchx84** [comment]
> Can we not replace it with git for the source instead of pypi, https://github.com/pybox2d/pybox2d/releases

We can try, but first, what is flatpak-pip-generator outputting ?

**Comment by sparshg** [comment]
> We can try, but first, what is flatpak-pip-generator outputting ?

It just outputs the previous version 2.3.2 which we are already on

**Comment by tchx84** [comment]
> > We can try, but first, what is flatpak-pip-generator outputting ?
> 
> It just outputs the previous version 2.3.2 which we are already on

Can you copy it here?

**Comment by sparshg** [comment]
`flatpak-pip-generator --runtime='org.gnome.Sdk//44' box2d --checker-data`

```json
{
    "name": "python3-box2d",
    "buildsystem": "simple",
    "build-commands": [
        "pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"box2d\" --no-build-isolation"
    ],
    "sources": [
        {
            "type": "file",
            "url": "https://files.pythonhosted.org/packages/cc/7b/ddb96fea1fa5b24f8929714ef483f64c33e9649e7aae066e5f5023ea426a/Box2D-2.3.2.tar.gz",
            "sha256": "d1557dffdf9c1d6c796ec5df53e3d93227bb026c14b8411d22c295edaa2fb225",
            "x-checker-data": {
                "type": "pypi",
                "name": "Box2D"
            }
        }
    ]
}
```

**Comment by tchx84** [comment]
> `flatpak-pip-generator --runtime='org.gnome.Sdk//44' box2d --checker-data`
> 
> ```json
> {
>     "name": "python3-box2d",
>     "buildsystem": "simple",
>     "build-commands": [
>         "pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"box2d\" --no-build-isolation"
>     ],
>     "sources": [
>         {
>             "type": "file",
>             "url": "https://files.pythonhosted.org/packages/cc/7b/ddb96fea1fa5b24f8929714ef483f64c33e9649e7aae066e5f5023ea426a/Box2D-2.3.2.tar.gz",
>             "sha256": "d1557dffdf9c1d6c796ec5df53e3d93227bb026c14b8411d22c295edaa2fb225",
>             "x-checker-data": {
>                 "type": "pypi",
>                 "name": "Box2D"
>             }
>         }
>     ]
> }
> ```

Ok, please explore other ways, but also ask in #flathub, because this should be working.. and not getting 

```
ERROR   src.manifest: Failed to check file box2d/Box2D-2.3.2.tar.gz with PyPIChecker: Couldn't find sdist for package Box2D: list index out of range
```

**Review by tchx84 — CHANGES_REQUESTED** [review]
Looks good, but let's clean up these commits a bit more. This git log should look like a live tutorial for the next person who updates it, so:

1. Make a separate commit for each module, e.g. `manifest: Update pygame`. You can piggy-back adding data checkers to the same commit, or make a separate commit for each module separately, e.g. `manifest: Apply flatpak-external-data-checker to pygame`.
2. Be consistent with how you write the commit messages, e.g. use uppercases consistently.
3. Add a README (similar to the one I wrote for the BaseApp) explaining how to build it locally and how to check for updates (using data checkers).

**Review by tchx84 — CHANGES_REQUESTED** [review]
I tested this, and found a few regressions, probably related to one of the dependencies upgrades.

Steps to reproduce one particular instance:

1. Install the latest "test" build from this PR, e.g. `flatpak install --user https://dl.flathub.org/build-repo/29498/org.sugarlabs.Pippy.flatpakref`
2. Run it, e.g. `flatpak run org.sugarlabs.Pippy//test`
3. Open the examples dialog.
4. Select "Graphics".
5. Select "Tree".
6. Run it.

You get:

Screenshot from 2023-06-13 11-22-16

If you repeat the same test with the stable version, e.g. 

```
flatpak install flathub org.sugarlabs.Pippy//stable
flatpak run org.sugarlabs.Pippy//stable
```

Screenshot from 2023-06-13 11-24-28

**Comment by tchx84** [comment]
> I tested this, and found a few regressions, probably related to one of the dependencies upgrades.
> 
> Steps to reproduce one particular instance:

@sparshg please NOTE that other examples suffer from similar regressions, please investigate further.

**Comment by sparshg** [comment]
I am unable to reproduce this issue, its working as expected for me...

**Comment by tchx84** [comment]
> I am unable to reproduce this issue, its working as expected for me...

Please reproduce this running on Fedora 38, try a VM.

**Comment by chimosky** [comment]
> I tested this, and found a few regressions, probably related to one of the dependencies upgrades.

Tested on F38 in a VM and I can reproduce.

**Comment by sparshg** [comment]
I have been testing this, the issue is related to pygame only as it is not embedding correctly. I did these tests:
- Pygame < 2, fails to build on 22.06, 23.06, with "Unable to find FreeType2 library"
- I tried upgrading all the dependencies except pygame, on BaseApp 21.04 and it worked fine.
- Then I upgraded pygame too, and I saw pygame didn't always embed correctly. It sometimes showed black screen, sometimes it was correct, sometimes it was windowed.
-  Upgraded BaseApp to 22.06, same pygame problem
- Upgraded to BaseApp 23.06, but now there is always black screen.
- Another test with other dependencies as it is but just pygame updated and BaseApp 23.06, same black screen.

**Comment by sparshg** [comment]
@tchx84 You tested it earlier right? Were the pygame examples running then? Because they should have never worked according to these tests I think

**Comment by chimosky** [comment]
> I have been testing this, the issue is related to pygame only as it is not embedding correctly. I did these tests:
> 
>     * Pygame < 2, fails to build on 22.06, 23.06, with "Unable to find FreeType2 library"
> 
>     * I tried upgrading all the dependencies except pygame, on BaseApp 21.04 and it worked fine.
> 
>     * Then I upgraded pygame too, and I saw pygame didn't always embed correctly. It sometimes showed black screen, sometimes it was correct, sometimes it was windowed.
> 
>     * Upgraded BaseApp to 22.06, same pygame problem
> 
>     * Upgraded to BaseApp 23.06, but now there is always black screen.
> 
>     * Another test with other dependencies as it is but just pygame updated and BaseApp 23.06, same black screen.

pygame 1.9 used SDL1 which had support for embedding windows with the `SDL_WINDOWID` and that was removed when pygame started using SDL2 but later added in pygame v2.3.0 if I remember correctly and you're using a higher version so I'm guessing your problem is related to that.

`SDL_WINDOWID` should still be supported so check if it exists and the value isn't None, that might be a good start.

**Comment by sparshg** [comment]
Okay, so I have figured out everything. `SDL_WINDOWID` is not used in Pippy since we aren't embedding pygame anywhere. Pygame just spawns a full screen window, its not related to flatpaks. You can reproduce the bug by running the following code simply with pygame (2.4.0, Fedora 48 Gnome), (btw it didn't happen on my mac). The issue is probably related to [this](https://github.com/pygame/pygame/issues/2538) and [this](https://github.com/pygame/pygame/issues/2631)

```py
import pygame
from pygame.locals import *
from random import *
import math

# always need to init first thing
pygame.init()

# create the window and keep track of the surface
# for drawing into
screen = pygame.display.set_mode((0, 0), pygame.FULLSCREEN)

# ask for screen's width and height
width, height = screen.get_size()
print(width, height)
# turn off the cursor
pygame.mouse.set_visible(False)

color = (250, 250, 250)
min_factor = 0.8
max_factor = 0.9
start_length = 130
min_length = 40
min_angle_delta = 0.4
max_angle_delta = 0.5

# start the screen all black
bgcolor = (0, 0, 0)
screen.fill(bgcolor)


def draw_tree(x, y, length, angle):
    x2 = x + length * math.sin(angle)
    y2 = y - length * math.cos(angle)
    pygame.draw.line(screen, color, (x, y), (x2, y2))

    if length > min_length:
        # draw left branch
        left_angle = angle - \
            uniform(min_angle_delta, max_angle_delta)
        left_length = length * \
            uniform(min_factor, max_factor)
        draw_tree(x2, y2, left_length, left_angle)
        # draw middle branch
        middle_length = length * \
            uniform(min_factor, max_factor)
        draw_tree(x2, y2, middle_length, angle)
        # draw right branch
        right_angle = angle + \
            uniform(min_angle_delta, max_angle_delta)
        right_length = length * \
            uniform(min_factor, max_factor)
        draw_tree(x2, y2, right_length, right_angle)

# clear the screen
screen.fill(bgcolor)

# draw a tree, starting at the bottom centre of the
# screen
draw_tree((width / 2), height - 20, start_length, 0)
pygame.display.flip()
while True:
    pass
```

The fix is simple enough, replace the line
`screen = pygame.display.set_mode((0, 0),  pygame.FULLSCREEN)`
with
`screen = pygame.display.set_mode((0, 0), pygame.NOFRAME | pygame.FULLSCREEN)`

Also this is only happening with the tree example not the snow example, probably because we are not calling the `pygame.display.flip()` every frame, if you add `draw_tree()` and `flip()` every frame, it works as well.

I think its best to include NOFRAME in this/every example to prevent this from happening again.

**Comment by chimosky** [comment]
@sparshg glad you found a possible solution, next step would be to open a PR upstream and test.

**Comment by sparshg** [comment]
@chimosky can you review https://github.com/sugarlabs/Pippy/pull/86

**Comment by sparshg** [comment]
Issue has been fixed in upstream, we can merge this once a new version has been released

**Review by tchx84 — APPROVED** [review]
well done!

**Pull request merged**